src/handler/sign_tx.c

```diff
@@ -1,103 +1,423 @@
-        G_context.req_type = CONFIRM_TRANSACTION;
-        G_context.state = STATE_NONE;
-
-        if (!buffer_read_u8(cdata, &G_context.bip32_path_len) ||
-            !buffer_read_bip32_path(cdata,
-                                    G_context.bip32_path,
-                                    (size_t) G_context.bip32_path_len)) {
-            return io_send_sw(SW_WRONG_DATA_LENGTH);
+
+bool verify_change_output(tx_output_t output, uint32_t *bip32_path, uint8_t bip32_len) {
+    uint8_t hash[20];
+    cx_ecfp_public_key_t public_key;
+    cx_ecfp_private_key_t private_key;
+    uint8_t chain_code[32];
+
+    derive_private_key(&private_key, chain_code, bip32_path, bip32_len);
+    init_public_key(&private_key, &public_key);
+    compress_public_key(public_key.W);
+    hash160(public_key.W, 33, hash);
+
+    // erase data
+    explicit_bzero(&private_key, sizeof(private_key));
+    explicit_bzero(&public_key, sizeof(public_key));
+    explicit_bzero(&chain_code, sizeof(chain_code));
+
+    // 0 means equals
+    return memcmp(hash, output.pubkey_hash, 20) == 0;
+}
+
+void read_change_output_info(buffer_t *cdata) {
+    uint8_t buffer[MAX_BIP32_PATH * 4 + 1] = {0};
+    uint8_t tmp;
+
+    // 1 byte for has_change_output and bip32 path len
+    if (!buffer_read_u8(cdata, &tmp)) THROW(SW_WRONG_DATA_LENGTH);
+
+    G_context.tx_info.has_change_output = (bool) tmp & 0x80;
+
+    if (G_context.tx_info.has_change_output) {
+        // 1 byte for change output index
+        if (!buffer_read_u8(cdata, &G_context.tx_info.change_output_index)) THROW(SW_WRONG_DATA_LENGTH);
+    }
+
+    buffer[0] = tmp & 0x0F;
+    if(buffer[0] > MAX_BIP32_PATH || !buffer_move(cdata, buffer+1, 4*buffer[0])) {
+        THROW(SW_WRONG_DATA_LENGTH);
+    }
+
+    buffer_t bufdata = {.ptr = buffer, .size=(size_t)(1+(4*buffer[0])), .offset = 0};
+
+    // buffer holds the serialized bip32 path that was read from cdata
+    if (!buffer_read_bip32_path(&bufdata, &G_context.tx_info.change_bip32_path)) THROW(SW_WRONG_DATA_LENGTH);
+}
+
+void read_tx_data(buffer_t *cdata) {
+    // // read version bytes (Big Endian)
+    // buffer_read_u16(cdata, G_context.tx_info.tx_version, BE);
+    // // read number of tokens, inputs and outputs (respectifully)
+    // buffer_read_u8(cdata, G_context.tx_info.remaining_tokens);
+    // buffer_read_u8(cdata, G_context.tx_info.remaining_inputs);
+    // buffer_read_u8(cdata, G_context.tx_info.outputs_len);
+
+    if (!(buffer_read_u16(cdata, &G_context.tx_info.tx_version, BE) && // read version bytes (Big Endian)
+        buffer_read_u8(cdata, &G_context.tx_info.remaining_tokens) && // read number of tokens, inputs and outputs, respectively
+        buffer_read_u8(cdata, &G_context.tx_info.remaining_inputs) &&
+        buffer_read_u8(cdata, &G_context.tx_info.outputs_len))) {
+        // if an error occurs reading
+        THROW(SW_WRONG_DATA_LENGTH);
+    }
+}
+
+void copy_sighash_all(buffer_t *cdata) {
```

Comment:
I guess the name is misleading. Why `copy_sighash_all`? Isn't it better to use `sighash_all_hash(...)`?